### ISSUE TITLE
Warning about compiler

### ISSUE THREAD

**Issue by jacob-carlborg** [MrcSnm/redub#14]
I tried Redub on one of my projects and got this warning:

```
Warning: The compiler 'dmd' that was specified in your system wasn't found. Redub found ldc2 and it will use for this compilation.
```

**Comment by p0nce** [comment]
Personally I always use this or that compiler, and would like to specify the default one using some sort of configuration.
Although it's true it's also up to me to make dmd a good target for debug builds.

**Comment by MrcSnm** [comment]
I've thought about 2 ways to solve that:

1. Create a local file which says which compiler you're using [ this will be done automatically after the first warning ]
2. Create a global file using the same approach

That means on the first run, it should warn, but after the first, it should change the compiler order it tries to find by making the last one found the first

**Comment by jacob-carlborg** [comment]
What does it warn about? What does "was specified in your system" mean? I haven't specified anything. I don't understand what needs to be solved.

**Comment by MrcSnm** [comment]
Redub now creates a file after the first default execution:

```json
{
    "defaultCompiler": "C:\\D\\dmd2\\windows\\bin64\\dmd.exe",
    "compilers": {
        "C:\\D\\dmd2\\windows\\bin64\\dmd.exe": [
            "dmd",
            "2.108.0",
            "2.108.0",
            "DMD64 D Compiler v2.108.0\r\nCopyright (C) 1999-2024 by The D Language Foundation, All Rights Reserved written by Walter Bright\r\n",
            638475961340000000
        ],
        "C:\\D\\ldc2\\ldc2-1.37.0-windows-multilib\\bin\\ldc2.exe": [
            "ldc2",
            "1.37.0",
            "2.107.1",
            "LDC - the LLVM D compiler (1.37.0):\n  based on DMD v2.107.1 and LLVM 17.0.6\n  built with LDC - the LLVM D compiler (1.37.0)\n  Default target: x86_64-pc-windows-msvc\n  Host CPU: znver4\n  http://dlang.org - http://wiki.dlang.org/LDC\n\n\n  Registered Targets:\n    aarch64     - AArch64 (little endian)\n    aarch64_32  - AArch64 (little endian ILP32)\n    aarch64_be  - AArch64 (big endian)\n    amdgcn      - AMD GCN GPUs\n    arm         - ARM\n    arm64       - ARM64 (little endian)\n    arm64_32    - ARM64 (little endian ILP32)\n    armeb       - ARM (big endian)\n    avr         - Atmel AVR Microcontroller\n    bpf         - BPF (host endian)\n    bpfeb       - BPF (big endian)\n    bpfel       - BPF (little endian)\n    hexagon     - Hexagon\n    lanai       - Lanai\n    loongarch32 - 32-bit LoongArch\n    loongarch64 - 64-bit LoongArch\n    mips        - MIPS (32-bit big endian)\n    mips64      - MIPS (64-bit big endian)\n    mips64el    - MIPS (64-bit little endian)\n    mipsel      - MIPS (32-bit little endian)\n    msp430      - MSP430 [experimental]\n    nvptx       - NVIDIA PTX 32-bit\n    nvptx64     - NVIDIA PTX 64-bit\n    ppc32       - PowerPC 32\n    ppc32le     - PowerPC 32 LE\n    ppc64       - PowerPC 64\n    ppc64le     - PowerPC 64 LE\n    r600        - AMD GPUs HD2XXX-HD6XXX\n    riscv32     - 32-bit RISC-V\n    riscv64     - 64-bit RISC-V\n    sparc       - Sparc\n    sparcel     - Sparc LE\n    sparcv9     - Sparc V9\n    spirv32     - SPIR-V 32-bit\n    spirv64     - SPIR-V 64-bit\n    systemz     - SystemZ\n    thumb       - Thumb\n    thumbeb     - Thumb (big endian)\n    ve          - VE\n    wasm32      - WebAssembly 32-bit\n    wasm64      - WebAssembly 64-bit\n    x86         - 32-bit X86: Pentium-Pro and above\n    x86-64      - 64-bit X86: EM64T and AMD64\n    xcore       - XCore\n",
            638450732882565945
        ]
    }
}
```

This file not only makes the compiler discovery faster since it becomes unnecessary to execute `compiler --version`, but also makes the annoying message disappear after the first run, since ldc2 was the first to be discovered on an exploratory run, it will be assigned as a default compiler and in the next run, it will be the first to try.

This update is available in 1.7.5

**Comment by jacob-carlborg** [comment]
Hmm, this looks like it's not compatible with having multiple versions of compilers installed, especially not using DVM [1]. DVM allows to run one compiler in one shell and a different compiler in a different shell. Here's a screenshot that shows how it's working:

<img width="920" alt="Screenshot 2024-07-19 at 08 22 31" src="https://github.com/user-attachments/assets/9058c1ce-8049-46fe-9538-124b973bc11c">

As you see, `which dmd` points to different paths. If Redub blindly reads the `defaultCompiler` property from the configuration file, it will probably find a compiler, but it might be using the wrong version. That's the worst behavior, no error but unexpected behavior, from the user's point of view.

[1] https://github.com/jacob-carlborg/dvm

**Comment by MrcSnm** [comment]
So, what happens is that in the first execution, it will try [dmd, ldc2]. After trying them both by searching them in PATH, it save the PATH compiler. The solution for that may be just not saving the defaultCompiler with the PATH already solved

**Comment by MrcSnm** [comment]
This issue were addressed in release v1.7.8
The new defaultCompiler property will only give the compiler name: This allows both taking the compiler from PATH and from the CWD.
Also, if you want to check the generated file, you can access the same path as from DUB's packages. You'll find it as redub_compilers.json

```json
{
    "defaultCompiler": "ldc2",
    "compilers": {
        "C:\\D\\dmd2\\windows\\bin64\\dmd.exe": [
            "dmd",
            "2.108.0",
            "2.108.0",
            "DMD64 D Compiler v2.108.0\r\nCopyright (C) 1999-2024 by The D Language Foundation, All Rights Reserved written by Walter Bright\r\n",
            638475961340000000
        ],
        "C:\\D\\ldc2\\ldc2-1.37.0-windows-multilib\\bin\\ldc2.exe": [
            "ldc2",
            "1.37.0",
            "2.107.1",
            "LDC - the LLVM D compiler (1.37.0):\n  based on DMD v2.107.1 and LLVM 17.0.6\n  built with LDC - the LLVM D compiler (1.37.0)\n  Default target: x86_64-pc-windows-msvc\n  Host CPU: znver4\n  http://dlang.org - http://wiki.dlang.org/LDC\n\n\n  Registered Targets:\n    aarch64     - AArch64 (little endian)\n    aarch64_32  - AArch64 (little endian ILP32)\n    aarch64_be  - AArch64 (big endian)\n    amdgcn      - AMD GCN GPUs\n    arm         - ARM\n    arm64       - ARM64 (little endian)\n    arm64_32    - ARM64 (little endian ILP32)\n    armeb       - ARM (big endian)\n    avr         - Atmel AVR Microcontroller\n    bpf         - BPF (host endian)\n    bpfeb       - BPF (big endian)\n    bpfel       - BPF (little endian)\n    hexagon     - Hexagon\n    lanai       - Lanai\n    loongarch32 - 32-bit LoongArch\n    loongarch64 - 64-bit LoongArch\n    mips        - MIPS (32-bit big endian)\n    mips64      - MIPS (64-bit big endian)\n    mips64el    - MIPS (64-bit little endian)\n    mipsel      - MIPS (32-bit little endian)\n    msp430      - MSP430 [experimental]\n    nvptx       - NVIDIA PTX 32-bit\n    nvptx64     - NVIDIA PTX 64-bit\n    ppc32       - PowerPC 32\n    ppc32le     - PowerPC 32 LE\n    ppc64       - PowerPC 64\n    ppc64le     - PowerPC 64 LE\n    r600        - AMD GPUs HD2XXX-HD6XXX\n    riscv32     - 32-bit RISC-V\n    riscv64     - 64-bit RISC-V\n    sparc       - Sparc\n    sparcel     - Sparc LE\n    sparcv9     - Sparc V9\n    spirv32     - SPIR-V 32-bit\n    spirv64     - SPIR-V 64-bit\n    systemz     - SystemZ\n    thumb       - Thumb\n    thumbeb     - Thumb (big endian)\n    ve          - VE\n    wasm32      - WebAssembly 32-bit\n    wasm64      - WebAssembly 64-bit\n    x86         - 32-bit X86: Pentium-Pro and above\n    x86-64      - 64-bit X86: EM64T and AMD64\n    xcore       - XCore\n",
            638450732882565945
        ]
    }
}
```

The new update also handles deleting the cache from older version so if there are future changes in the JSON, you won't need to deal it yourself
